crates/typst/src/foundations/calc.rs

```diff
@@ -909,29 +911,61 @@ pub fn quo(
     Ok(floor(dividend.apply2(divisor.v, Div::div, Div::div)))
 }
 
-/// Calculates the euclidean norm of a sequence of numbers.
+/// Calculates the euclidean norm of a sequence of values.
 ///
 /// ```example
 /// #calc.norm(1, 2, -3, 0.5)
+/// #calc.norm(1in, 2cm)
+/// #calc.norm(3em, 4em)
 /// ```
 #[func]
 pub fn norm(
-    /// The callsite span.
-    span: Span,
-    /// The sequence of numbers from which to calculate the norm.
+    /// The sequence of values from which to calculate the norm.
     /// Must not be empty.
     #[variadic]
-    values: Vec<Spanned<Num>>,
-) -> SourceResult<f64> {
-    let mut sum = 0f64;
-    if values.is_empty() {
-        bail!(span, "expected at least one value");
-    }
-
-    for Spanned { v, span: _ } in values {
-        sum += v.float().powi(2)
+    values: Vec<Spanned<Value>>,
+) -> SourceResult<Value> {
+    let mut sum = 0.0;
+    if let Some(Spanned { v, span }) = values.first() {
+        match v {
+            Value::Int(_) | Value::Float(_) => {
+                for Spanned { v, span } in values {
+                    match v {
+                        Value::Int(n) => sum += (n as f64).powi(2),
+                        Value::Float(n) => sum += n.powi(2),
+                        _ => bail!(span, "expected a number"),
+                    }
+                }
+                Ok(Value::Float(sum.sqrt()))
+            }
+            Value::Length(Length { em, .. }) if em.is_zero() => {
```

Comment:
This function could give a hint to call `to-absolute` when it errors from mixing absolute and em units.

Reply:
Sorry, I'm not very familiar with the code base and not sure how to do it. Could you explain it?

Reply:
Hints can be given through the `bail!` macro. You could search its definition and see the document for developers, which should help you do this.

Reply:
> Hints can be given through the `bail!` macro. You could search its definition and see the document for developers, which should help you do this.

Got it, thanks!